src/capped_hashmap.rs

```diff
@@ -0,0 +1,172 @@
+use std::{
+    cmp::Eq,
+    collections::{HashMap, VecDeque},
+    hash::Hash,
+};
+
+pub struct CappedHashMap<K, V>
+where
+    K: Hash + Eq + Copy + Clone,
+{
+    capacity: usize,
+    inner: HashMap<K, V>,
+    last_items: VecDeque<K>,
+}
+
+impl<K, V> CappedHashMap<K, V>
+where
+    K: Hash + Eq + Copy + Clone,
+{
+    pub fn new(capacity: usize) -> Self {
+        Self {
+            // We add a little bit margin to the capacity, because we always insert first and then check
+            // if collection is full. So we don't want the hashmap to double it's capacity because of that.
+            capacity: capacity + 1,
+            inner: HashMap::with_capacity(capacity),
+            last_items: VecDeque::with_capacity(capacity),
+        }
+    }
+
+    /// Inserts an new item to the collection. Return Some(key) where key is the
+    /// key that was removed when we reach the max capacity. Otherwise returns None.
+    pub fn insert(&mut self, k: K, v: V) -> Option<K> {
+        let mut ret = None;
+
+        // replacing a value should not push any new items to last_items
+        if let None = self.inner.insert(k, v) {
+            self.last_items.push_front(k);
+        }
+
+        if self.last_items.len() > self.capacity - 1 {
```

Comment:
I think we should have logs when we reach a quarter of the capacity, half of the capacity, 90% of the capacity, or something like that (so that we now something is happening and unfinished signatures are piling up

Reply:
also imagine `capacity` is set to 5, this means that the hashmap and vecdeque both have capacity 4. So when the insertion above makes `inner` and `last_items` full, the check here will be `4 > 4 = false` and do nothing. This means that the next insert will resize both the hashmap and vecdeque. I think you made a mistake in the initialization, you should keep the capacity, but have the two structures have capacity + 1 instead. Maybe a test checking for the array capacity would help with making sure that the logic works :)

Reply:
>  So when the insertion above makes inner and last_items full, the check here will be 4 > 4 = false and do nothing. Yes that is correct. If the number of items are 4 then we don't want to remove the oldest one. This is intentional. We want to remove when we reach the full capacity which is 5, in this example.

Reply:
I've read about the capacity and how it changes. There is a good explanation [here](https://stackoverflow.com/questions/76114698/capacity-of-hashmaps-in-rust-rust-by-practice/76114888#76114888).

I also tested it and it looks like it can double even though the length remains the same.

> it has to reserve "slack space", the amount of which depends on the implementation of the hashmap (and specifically its collision resolution algorithm).

I guess the more keys we insert the highest the chance of a collision (even though some keys are removed) so it looks like it follows a conservative approach by doubling the capacity. 

I believe this is not a big deal neither a huge performance implication given that we're not gonna be storing several thousand or millions of entries in the hash table. And besides capacity change does not mean entries are moved to a new memory location.